lib/usagereporter/teleport/aggregating/submitter.go

```diff
@@ -0,0 +1,192 @@
+		return trace.BadParameter("missing Backend")
+	}
+	if cfg.Log == nil {
+		return trace.BadParameter("missing Log")
+	}
+	if cfg.Status == nil {
+		return trace.BadParameter("missing Status")
+	}
+	if cfg.Submitter == nil {
+		return trace.BadParameter("missing Submitter")
+	}
+	return nil
+}
+
+// RunSubmitter periodically fetches aggregated reports from the backend and
+// sends them in small batches with the provided submitter, before deleting
+// them. Must only be called after validating the config object with
+// CheckAndSetDefaults, and should probably be called in a goroutine.
+func RunSubmitter(ctx context.Context, cfg SubmitterConfig) {
+	iv := interval.New(interval.Config{
+		FirstDuration: utils.HalfJitter(2 * submitInterval),
+		Duration:      submitInterval,
+		Jitter:        retryutils.NewSeventhJitter(),
+	})
+	defer iv.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-iv.Next():
+		}
+
+		submitOnce(ctx, cfg)
+	}
+}
+
+func submitOnce(ctx context.Context, c SubmitterConfig) {
+	svc := reportService{c.Backend}
+
+	reports, err := svc.listUserActivityReports(ctx, submitBatchSize)
+	if err != nil {
+		c.Log.WithError(err).Error("Failed to load usage reports for submission.")
+		return
+	}
+
+	if len(reports) < 1 {
+		if _, err := c.Status.GetClusterAlerts(ctx, types.GetClusterAlertsRequest{
+			AlertID: alertName,
+		}); err != nil && trace.IsNotFound(err) {
+			// if we can confirm that there's no cluster alert we go ahead and
+			// exit here without attempting the delete (reads are cheaper than
+			// writes)
+			return
+		}
+		err := c.Status.DeleteClusterAlert(ctx, alertName)
+		if err == nil {
+			c.Log.Infof("Deleted cluster alert %v after successfully clearing usage report backlog.", alertName)
+		} else if !trace.IsNotFound(err) {
+			c.Log.WithError(err).Errorf("Failed to delete cluster alert %v.", alertName)
+		}
+		return
+	}
+
+	if err := svc.createUserActivityReportsLock(ctx, submitLockDuration); err != nil {
+		if trace.IsAlreadyExists(err) {
+			c.Log.Debugf("Failed to acquire lock %v, already held.", userActivityReportsLock)
+		} else {
+			c.Log.WithError(err).Errorf("Failed to acquire lock %v.", userActivityReportsLock)
+		}
+		return
+	}
+
+	lockCtx, cancel := context.WithTimeout(ctx, submitLockDuration*9/10)
+	defer cancel()
+
+	batchUUID, err := c.Submitter(lockCtx, &prehogv1.SubmitUsageReportsRequest{
+		UserActivity: reports,
+	})
+	if err != nil {
+		c.Log.WithError(err).WithFields(logrus.Fields{
+			"reports":       len(reports),
+			"oldest_report": reports[0].GetStartTime().AsTime(),
+			"newest_report": reports[len(reports)-1].GetStartTime().AsTime(),
+		}).Error("Failed to send usage reports.")
+
+		if time.Since(reports[0].StartTime.AsTime()) <= alertGraceDuration {
+			return
+		}
+		alert, err := types.NewClusterAlert(
+			alertName,
+			alertMessage,
+			types.WithAlertLabel(types.AlertOnLogin, "yes"),
+			types.WithAlertLabel(types.AlertPermitAll, "yes"),
```

Comment:
I went with the same alert visibility as the license warnings (albeit at `AlertSeverity_MEDIUM` rather than `HIGH`, so orange rather than red); should I reduce the visibility? I think the intention is to make it similar to the old-world Houston license warning, which - from what I can tell - was shown to everyone.

Reply:
FWIW nowadays we have alert acknowledgement in place, so admins can just hide the alert once they are aware of the problem and are attempting to fix it.